erpnext/crm/doctype/email_campaign/email_campaign.py

```diff
@@ -0,0 +1,102 @@
+		for entry in campaign.get("campaign_schedules"):
+			send_after_days.append(entry.send_after_days)
+		try:
+			end_date = add_days(getdate(self.start_date), max(send_after_days))
+		except ValueError:
+			frappe.throw(_("Please set up the Campaign Schedule in the Campaign {0}").format(self.campaign_name))
+
+	def validate_lead(self):
+		lead_email_id = frappe.db.get_value("Lead", self.recipient, 'email_id')
+		if not lead_email_id:
+			lead_name = frappe.db.get_value("Lead", self.recipient, 'lead_name')
+			frappe.throw(_("Please set an email id for the Lead {0}").format(lead_name))
+
+	def validate_email_campaign_already_exists(self):
+		email_campaign_exists = frappe.db.exists("Email Campaign", {
+			"campaign_name": self.campaign_name,
+			"recipient": self.recipient,
+			"status": ("in", ["In Progress", "Scheduled"])
+		})
+		if email_campaign_exists:
+			frappe.throw(_("The Campaign '{0}' already exists for the {1} '{2}'").format(self.campaign_name, self.email_campaign_for, self.recipient))
+
+	def update_status(self):
+		start_date = getdate(self.start_date)
+		end_date = getdate(self.end_date)
+		today_date = getdate(today())
+		if start_date > today_date:
+			self.status = "Scheduled"
+		elif end_date >= today_date:
+			self.status = "In Progress"
+		elif end_date < today_date:
+			self.status = "Completed"
+
+#called through hooks to send campaign mails to leads
+def send_email_to_leads_or_contacts():
+	email_campaigns = frappe.get_all("Email Campaign", filters = { 'status': ('not in', ['Unsubscribed', 'Completed', 'Scheduled']) })
+	for camp in email_campaigns:
+		email_campaign = frappe.get_doc("Email Campaign", camp.name)
+		campaign = frappe.get_cached_doc("Campaign", email_campaign.campaign_name)
+		for entry in campaign.get("campaign_schedules"):
+			scheduled_date = add_days(email_campaign.get('start_date'), entry.get('send_after_days'))
+			if scheduled_date == getdate(today()):
+				send_mail(entry, email_campaign)
+
+def send_mail(entry, email_campaign):
+	recipient = frappe.db.get_value(email_campaign.email_campaign_for, email_campaign.get("recipient"), 'email_id')
+
+	email_template = frappe.get_doc("Email Template", entry.get("email_template"))
+	sender = frappe.db.get_value("User", email_campaign.get("sender"), 'email')
+
+	# send mail and link communication to document
+	comm = make(
+		doctype = "Email Campaign",
+		name = email_campaign.name,
+		subject = email_template.get("subject"),
+		content = email_template.get("response"),
+		sender = sender,
+		recipients = recipient,
+		communication_medium = "Email",
+		sent_or_received = "Sent",
+		send_email = True,
+		email_template = email_template.name
+	)
+	return comm
+
+#called from hooks on doc_event Email Unsubscribe
+def unsubscribe_recipient(unsubscribe, method):
+	if unsubscribe.reference_doctype == 'Email Campaign':
+		frappe.db.set_value("Email Campaign", unsubscribe.reference_name, "status", "Unsubscribed")
```

Comment:
@ruchamahabal I know this was a long time back, but IMO this makes no sense at all. (Marking the whole **Email Campaign** aus "Unsubscribed", just because a single recipient has unsubscribed. Which will then prevent status updates below.)

Reply:
Umm Email Campaign and Recipient have a 1-1 relationship. So its specific to a recipient IIRC

Reply:
You're right, my bad. Originally it was a 1-1 relationship, but since https://github.com/frappe/erpnext/pull/22731 it can be a 1-n relationship.